sycl/doc/extensions/Matrix/dpcpp-joint-matrix.asciidoc

```diff
@@ -0,0 +1,245 @@
+struct matrix;
+enum class matrix_layout {
+  row_major,
+  col_major,
+  packed_a,
+  packed_b
+};
+}
+```
+
+
+### Memory Scope
+For the memory scope, we have two solutions. The long term solution is to use the proposed https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/LocalMemory/SYCL_INTEL_local_memory.asciidoc[`group_local_memory` extension] to allocate the matrix in local memory associated with a SYCL group as shown in the example below.
+
+```c++
+multi_ptr<matrix<T>, address_space::local_space> tA_ptr = group_local_memory<matrix<sub_group, int8_t, tM, tN>>(sg);
+```
+However, sub-group local memory is not yet well defined in {dpcpp}. Moreover, the representation of this notion in LLVM IR and SPIR-V is not yet clear. Hence, for this proposal, we use `joint_matrix` instead of `matrix` to emphasis that the matrix is shared among a group of work items and is not private to each work item. Moreover, the memory scope is added as an additional constructor argument. This results into the following description:
+
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+template <typename Group, typename T, size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent, matrix_layout l = row_major>
+struct joint_matrix {
+    joint_matrix(Group g) {}
+};
+}
+```
+When the group is a `sycl::sub_group`, a matrix is declared as follows:
+
+```c++
+joint_matrix<sub_group, int8_t, tM, tN> tA(sg); 
+```
+
+
+## Matrix Operations and their Execution Scope
+We define three new functions needed to perform the main and common operations on matrices namely, load, store, and the actual multiply and add operation. This set of functions can be easily extended if the tensor hardware implements new features.
+
+The base pointer determines the starting address of the matrix to be loaded/stored. `layout` determines whether the data are being read/written in a row (`row_major`), column major (`column_major`) fashion, or if the data has already been transformed into VNNI format (`packed_a`, `packed_b`). 
+
+Since the matrix functions are group operations (as defined in Section 4.17.3 of the SYCL specification), the matrix API has to be accessed by all the work-items in the group in a non-diverged control flow. The `Group` template argument can be a work-group or a sub-group. These functions will be called once by all the work items in a group.
+
+To be aligned with the SYCL 2020 group algorithms, an additional group argument is added to the matrix operations to designate that these functions are collective operations. The {dpcpp} syntax is the following: 
+
+```c++
+namespace sycl::ext::intel::experimental::matrix {
+template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          access::address_space Space>
+  void joint_matrix_load(Group sg, joint_matrix<Group, T, NumRows, NumCols> &res,
+		    multi_ptr<T, Space> src, size_t stride, matrix_layout l = row_major);
```

Comment:
Yes we need both. layout in joint_matrix specifies the layout of the matrix. Matrix is in a different storage from the memory that contained the original data and where we are doing the load from. We need layout for each.

I don't think layout should be a template parameter on load/store because as of now the user does not need to specify any template argument, they are all deduced. If we add layout as template parameter, the user will have then to specify all the template argument which can be a lot. Moreover, it will not be consistent with the current DPC++ joint free functions.

OK for the indentation.

Reply:
> Yes we need both. layout in joint_matrix specifies the layout of the matrix. Matrix is in a different storage from the memory that contained the original data and where we are doing the load from. We need layout for each.

Oh, so each matrix can have two independent layouts?  One specifies the layout of the matrix in memory, and the other specifies the layout of the matrix after it has been loaded into the hardware register?  This seems surprisingly weird to me, and it is not how I thought it worked from reading the extension API.  If this is how it works, I think this needs to be described in more detail in the API spec.

FWIW, I assumed the matrix had some fixed layout after it was loaded into the hardware register, and I assumed the only layout that the user needed to worry about was the layout of the matrix in memory.

> I don't think layout should be a template parameter on load/store because as of now the user does not need to specify any template argument, they are all deduced. If we add layout as template parameter, the user will have then to specify all the template argument which can be a lot.

I believe the API you have now (without the extra `Layout` template parameter) won't compile in the case when the `joint_matrix` parameter has any layout other than the default of `row_major`.  Therefore, I think you need to add the `Layout` template parameter in order for the API to work.  Here is a mock-up of your existing API, which demonstrates the problem:

```
#include <cstddef>

namespace sycl {
namespace access {

enum class address_space : int {
  global_space,
  local_space,
  constant_space, // Deprecated in SYCL 2020
  private_space,
  generic_space,
};

enum class decorated : int {
  no,
  yes,
  legacy,
};

} // namespace access

inline constexpr std::size_t dynamic_extent = -1;
template <typename ElementType, access::address_space Space,
          access::decorated DecorateAddress>
class multi_ptr{};
class sub_group{};

} // namespace sycl


namespace sycl::ext::intel::experimental::matrix {

enum class matrix_layout {
  row_major,
  col_major,
  packed_a,
  packed_b
};

template <typename Group, typename T, size_t Rows=sycl::dynamic_extent,
          size_t Cols=sycl::dynamic_extent,
          matrix_layout Layout = matrix_layout::row_major>
struct joint_matrix {
    joint_matrix(Group g) {}
};

template <typename Group, typename T, size_t NumRows, size_t NumCols,
          access::address_space Space, access::decorated IsDecorated>
void joint_matrix_load(Group sg, joint_matrix<Group, T, NumRows, NumCols> &res,
                       multi_ptr<T, Space, IsDecorated> src, size_t stride,
                       matrix_layout layout = matrix_layout::row_major);

} // namespace sycl::ext::intel::experimental::matrix


namespace syclmatrix = sycl::ext::intel::experimental::matrix;

int main() {
  sycl::sub_group sg;
  sycl::multi_ptr<int, sycl::access::address_space::global_space, sycl::access::decorated::no>  mp;
  syclmatrix::joint_matrix<sycl::sub_group, int, 4, 4, syclmatrix::matrix_layout::packed_b> tB(sg);
  syclmatrix::joint_matrix_load(sg, tB, mp, 4);
}
```

Compiling that mockup gives an error:

```
$ clang++ -std=c++17 -pedantic -c matrix-layout.cpp 
matrix-layout.cpp:62:3: error: no matching function for call to 'joint_matrix_load'
  syclmatrix::joint_matrix_load(sg, tB, mp, 4);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
matrix-layout.cpp:49:6: note: candidate template ignored: could not match 0 against 3
void joint_matrix_load(Group sg, joint_matrix<Group, T, NumRows, NumCols> &res,
     ^
```

This can be fixed by adding the `Layout` template parameter as I note above.  Also notice that the compiler can still deduce the template parameters to `joint_matrix_load`.

```
template <typename Group, typename T, size_t NumRows, size_t NumCols,
          access::address_space Space, access::decorated IsDecorated, matrix_layout Layout>
void joint_matrix_load(Group sg, joint_matrix<Group, T, NumRows, NumCols, Layout> &res,
                       multi_ptr<T, Space, IsDecorated> src, size_t stride,
                       matrix_layout layout = matrix_layout::row_major);
```

BTW, notice that I also added a template parameter `IsDecorated`.  That's needed too in order for the `multi_ptr` parameter to work.  Of course, you can decide to change the order of the template parameters, but I think you need all the ones I list above.

Reply:
>> FWIW, I assumed the matrix had some fixed layout after it was loaded into the hardware register, and I assumed the only layout that the user needed to worry about was the layout of the matrix in memory.

For our hardware, you are right, the matrix is always packed (VNNIed). For performance, there is only one layout that matters for our hardware currently. 
Now, what does it mean for B to be packed (packed_b):
If the regular matrix shape is B[K][N] and type is bf16, the VNNI shape is B[K/2][N*2].
We don’t want the user to worry about converting those sizes. So the way to write this is: joint_matrix<bf16,K,N, packed_b> B; This is also telling the implementation that this one is the matrix B, the one that needs VNNIing. 

So having layout on matrix B is necessary to allocate the right sizes for the AMX tile.


Also, this interface is oneAPI interface, it is supposed to handle other hardware too. For instance, in the nvidia case, there is flexibility for the matrix to have row or column major layout. So for generality, the matrix itself can have any layout.

Reply:
>> This can be fixed by adding the Layout template parameter as I note above. 

You are right, sorry for not spotting this earlier

Reply:
> > > This can be fixed by adding the Layout template parameter as I note above.
> 
> @gmlueck  You are right, sorry for not spotting this earlier. I will update the arguments for these 3 functions

Reply:
Regarding the layout, I think the spec could clarify these points, which seemed confusing to me:

* What is the difference between `packed_a` and `packed_b` formats.
* What is the difference between the `Layout` template parameter and the `layout` runtime parameter in `joint_matrix_{load,store}()`.  When does the application use one vs. the other.